include/rawpheno.download.form.inc

```diff
@@ -152,7 +167,35 @@ function rawpheno_download($form, &$form_state) {
   );
 
   // Get traits given a location and project.
-  $opt_trait = rawpheno_download_load_traits($location_id, $project_id);
+  if (isset($project_selected) && isset($location)) {
+    $project_id = $project_selected;
+    $location_id = $location;
+  }
+
+  // Extract list of unique traits from rawdata summary materialized view.
+  // Filtered by location and project id number. Default to all location in the default project.
+  $sql = "SELECT ARRAY_TO_STRING(ARRAY_AGG(DISTINCT all_traits), ',') AS all_traits
+          FROM {rawpheno_rawdata_mview}
+          WHERE location IN(:location) AND plant_id IN
+            (SELECT plant_id FROM pheno_plant_project WHERE project_id = :project_id)
+          LIMIT 1";
+
+  $traits = chado_query($sql, array(':location' => $location_id, ':project_id' => $project_id))
+    ->fetchObject();
+
+  $temp_traits = str_getcsv($traits->all_traits, ',');
+  $traits_array = array_unique($temp_traits);
+
+  // Traits are all in cvterm_ids, we need the human-readable.
+  // Note: query above may return duplicate trait ids since multiple locations share same traits
+  //       but this query ensures that traits are unique.
+  $sql = "SELECT cvterm_id, name
+          FROM {cvterm}
+          WHERE cvterm_id in (:traits) AND name NOT IN ('Rep', 'Entry', 'Location', 'Name', 'Plot', 'Planting Date (date)', '# of Seeds Planted (count)')
+          GROUP BY cvterm_id, name ORDER BY name ASC";
+
+  $opt_trait = chado_query($sql, array(':traits' => $traits_array))

```

Comment:
Is there a reason for doing this in two queries? It would be faster to create a json object above with both the id and the name for each trait.